File: inventory-framework-platform-bukkit/src/main/java/me/devnatan/inventoryframework/context/SlotContext.java
package me.devnatan.inventoryframework.context;

import java.util.List;
import java.util.Map;
import java.util.UUID;
import lombok.ToString;
import me.devnatan.inventoryframework.RootView;
import me.devnatan.inventoryframework.View;
import me.devnatan.inventoryframework.ViewConfig;
import me.devnatan.inventoryframework.ViewContainer;
import me.devnatan.inventoryframework.Viewer;
import me.devnatan.inventoryframework.pipeline.Pipeline;
import me.devnatan.inventoryframework.state.StateValue;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

@ToString(onlyExplicitlyIncluded = true)
public abstract class SlotContext extends PlatformContext implements IFSlotContext, Context {

    // --- Inherited ---
    private final IFRenderContext parent;

    // --- Properties ---
    @ToString.Include
    private int slot;

    protected SlotContext(int slot, @NotNull IFRenderContext parent) {
        this.slot = slot;
        this.parent = parent;
    }

    public abstract ItemStack getItem();

    @Override
    public final @NotNull RenderContext getParent() {
        return (RenderContext) parent;
    }

    @Override
    public final int getSlot() {
        return slot;
    }

    @Override
    public final void setSlot(int slot) {
        this.slot = slot;
    }

    @Override
    public @NotNull Pipeline<IFContext> getPipeline() {
        return getParent().getPipeline();
    }

    @Override
    public final @NotNull Map<String, Viewer> getIndexedViewers() {
        return getParent().getIndexedViewers();
    }

    @Override
    public final @NotNull String getTitle() {
        return getParent().getTitle();
    }

    @Override
    public final void update() {
        getParent().update();
    }

    @Override
    public final @NotNull UUID getId() {
        return getParent().getId();
    }

    @Override
    public final @NotNull ViewConfig getConfig() {
        return getParent().getConfig();
    }

    @Override
    public final @NotNull ViewContainer getContainer() {
        return getParent().getContainer();
    }

    @Override
    public final @NotNull View getRoot() {
        return getParent().getRoot();
    }

    @Override
    public final Object getInitialData() {
        return getParent().getInitialData();
    }

    @Override
    public void setInitialData(Object initialData) {
        getParent().setInitialData(initialData);
    }

    @Override
    public List<Player> getAllPlayers() {
        return getParent().getAllPlayers();
    }

    @Override
    public void updateTitleForPlayer(@NotNull String title, @NotNull Player player) {
        getParent().updateTitleForPlayer(title, player);
    }

    @Override
    public void resetTitleForPlayer(@NotNull Player player) {
        getParent().resetTitleForPlayer(player);
    }

    @Override
    public final boolean isActive() {
        return getParent().isActive();
    }

    @Override
    public final void setActive(boolean active) {
        getParent().setActive(active);
    }

    @Override
    public final boolean isEndless() {
        return getParent().isEndless();
    }

    @Override
    public final void setEndless(boolean endless) {
        getParent().setEndless(endless);
    }

    @Override
    public final void back() {
        getParent().back();
    }

    @Override
    public final void back(Object initialData) {
        getParent().back(initialData);
    }

    @Override
    public final boolean canBack() {
        return getParent().canBack();
    }

    @Override
    public final void closeForPlayer() {
        getParent().closeForPlayer();
    }

    @Override
    public final void simulateCloseForPlayer() {
        getParent().simulateCloseForPlayer();
    }

    @Override
    public final void openForPlayer(@NotNull Class<? extends RootView> other) {
        getParent().openForPlayer(other);
    }

    @Override
    public final void openForPlayer(@NotNull Class<? extends RootView> other, Object initialData) {
        getParent().openForPlayer(other, initialData);
    }

    @Override
    public final void updateTitleForPlayer(@NotNull String title) {
        getParent().updateTitleForPlayer(title);
    }

    @Override
    public final void resetTitleForPlayer() {
        getParent().resetTitleForPlayer();
    }

    @Override
    public final Map<Long, StateValue> getStateValues() {
        return getParent().getStateValues();
    }

    @Override
    protected final void setUpdatedTitle(String updatedTitle) {
        getParent().setUpdatedTitle(updatedTitle);
    }
}

File: inventory-framework-platform-minestom/src/main/kotlin/me/devnatan/inventoryframework/context/SlotContext.kt
package me.devnatan.inventoryframework.context

import me.devnatan.inventoryframework.View
import me.devnatan.inventoryframework.ViewConfig
import me.devnatan.inventoryframework.ViewContainer
import me.devnatan.inventoryframework.Viewer
import me.devnatan.inventoryframework.component.Component
import me.devnatan.inventoryframework.state.State
import me.devnatan.inventoryframework.state.StateValue
import me.devnatan.inventoryframework.state.StateWatcher
import net.minestom.server.entity.Player
import net.minestom.server.item.ItemStack
import org.jetbrains.annotations.ApiStatus
import org.jetbrains.annotations.UnmodifiableView
import java.util.UUID

abstract class SlotContext
    @ApiStatus.Internal
    protected constructor(
        private var slot: Int,
        private val parent: IFRenderContext,
    ) : PlatformContext(),
        IFSlotContext,
        Context {
        abstract val item: ItemStack

        override fun getParent(): RenderContext = parent as RenderContext

        override fun getSlot(): Int = slot

        override fun setSlot(slot: Int) {
            this.slot = slot
        }

        override fun getIndexedViewers(): Map<String, Viewer> = getParent().indexedViewers

        override fun getTitle(): String = getParent().title

        override fun getComponents(): @UnmodifiableView MutableList<Component> = getParent().components

        override fun getInternalComponents(): List<Component> = getParent().internalComponents

        override fun getComponentsAt(position: Int): List<Component> = getParent().getComponentsAt(position)

        override fun addComponent(component: Component) {
            getParent().addComponent(component)
        }

        override fun removeComponent(component: Component) {
            getParent().removeComponent(component)
        }

        override fun renderComponent(component: Component) {
            getParent().renderComponent(component)
        }

        override fun updateComponent(
            component: Component,
            force: Boolean,
        ) {
            getParent().updateComponent(component, force)
        }

        override fun performClickInComponent(
            component: Component,
            viewer: Viewer,
            clickedContainer: ViewContainer,
            platformEvent: Any,
            clickedSlot: Int,
            combined: Boolean,
        ) {
            getParent()
                .performClickInComponent(
                    component,
                    viewer,
                    clickedContainer,
                    platformEvent,
                    clickedSlot,
                    combined,
                )
        }

        override fun update() {
            getParent().update()
        }

        override fun getRawStateValue(state: State<*>?): Any = getParent().getRawStateValue(state)

        override fun getInternalStateValue(state: State<*>): StateValue = getParent().getInternalStateValue(state)

        override fun getUninitializedStateValue(stateId: Long): StateValue = getParent().getUninitializedStateValue(stateId)

        override fun initializeState(
            id: Long,
            value: StateValue,
        ) {
            getParent().initializeState(id, value)
        }

        override fun updateState(
            state: State<*>,
            value: Any,
        ) {
            getParent().updateState(state, value)
        }

        override fun watchState(
            id: Long,
            listener: StateWatcher,
        ) {
            getParent().watchState(id, listener)
        }

        override fun getId(): UUID = getParent().id

        override fun getConfig(): ViewConfig = getParent().config

        override fun getContainer(): ViewContainer = getParent().container

        override fun getRoot(): View = getParent().getRoot()

        override fun getInitialData(): Any = getParent().initialData

        override fun setInitialData(initialData: Any) {
            getParent().initialData = initialData
        }

        override val allPlayers: List<Player>
            get() = getParent().allPlayers

        override fun updateTitleForPlayer(
            title: net.kyori.adventure.text.Component,
            player: Player,
        ) {
            getParent().updateTitleForPlayer(title, player)
        }

        override fun resetTitleForPlayer(player: Player) {
            getParent().resetTitleForPlayer(player)
        }

        override fun isActive(): Boolean = getParent().isActive

        override fun setActive(active: Boolean) {
            getParent().isActive = active
        }

        override fun isEndless(): Boolean = getParent().isEndless

        override fun setEndless(endless: Boolean) {
            getParent().isEndless = endless
        }

        override fun back() {
            getParent().back()
        }

        override fun back(initialData: Any) {
            getParent().back(initialData)
        }

        override fun canBack(): Boolean = getParent().canBack()
    }
